### ISSUE TITLE
Kimi K3 TP8/EP8 b8 roofline on MI350X: from 224 toward 400–500 decode tok/s

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#58]
# Kimi K3 TP8/EP8 b8 roofline on MI350X: from 224 toward 400–500 decode tok/s

## Executive answer

For graph-replayed Kimi K3 decode at b8/c8, 4k/1k on eight MI350X GPUs, the **hardware-only HBM roof is approximately 1,000–1,120 aggregate decode tok/s on the critical EP8 rank** (7.1–8.0 ms per eight-token model call). A perfectly balanced average rank reaches about 1,300 tok/s, but that is not the system roof: every one of the 92 MoE layers waits for its slowest rank.

Today TokenSpeed TP8/EP8 measures **35.77 ms TPOT**, equivalent to **223.6 steady-state decode tok/s**, and 207.05 end-to-end output tok/s after prefill/scheduling overhead. SGLang TP8/EP1 measures 25.68 ms or 311.5 steady-state decode tok/s and 295.31 output tok/s. The practical progression should therefore be:

1. **Match:** ≤26.7 ms / ≥300 decode tok/s.
2. **Production:** ≤20 ms / ≥400 decode tok/s.
3. **Stretch:** ≤16 ms / ≥500 decode tok/s.

The hardware-only roof is deliberately not a release target. Reaching 400–500 tok/s already requires substantially better grouped-expert bandwidth, fixed-weight reuse, rank balance, collective arrival, and device scheduling.

This issue covers steady-state decode. TTFT/prefill has a separate compute and collective roofline.

## Hardware and model inputs

AMD's [MI350X specification](https://www.amd.com/en/products/accelerators/instinct/mi350/mi350x.html) and [8-GPU platform specification](https://www.amd.com/en/products/accelerators/instinct/mi350/mi350x/platform.html) provide the hardware ceilings:

- 8 TB/s HBM3E bandwidth and 288 GB HBM per GPU.
- 9.2 PFLOP/s MXFP4 and 2.3 PFLOP/s BF16 matrix peak per GPU.
- 256 MB last-level cache per GPU, far below a decode step's weight footprint.
- Seven 153 GB/s Infinity Fabric links per GPU. K3's per-layer reductions are generally rendezvous/latency-bound rather than limited by bulk link bandwidth.

K3's [published configuration](https://huggingface.co/moonshotai/Kimi-K3/blob/main/config.json) and the real local checkpoint give:

- 93 layers: 69 KDA, 24 MLA, one dense layer and 92 MoE layers.
- Hidden size 7,168; routed latent size 3,584; expert intermediate size 3,072.
- 896 routed experts, top-16 per token, and two shared experts.
- Routed experts use MXFP4. Attention, router, latent projections and shared experts remain BF16 in this checkpoint.
- TP8 shards tensor-parallel matrices. EP8 places 112 routed experts on each GPU.

## Why b8 has a different roof from b1

The fixed attention/router/projection weights can ideally be loaded once and reused across all eight rows of an M=8 GEMM. Routed-expert weights do not receive the same 8× reuse: eight tokens create 128 expert assignments, and collisions are uncommon across 896 experts.

Under a simple uniform-routing model, the expected number of distinct experts touched on each rank is:

```text
local experts = 112
probability one expert is selected by one token = 16 / 896

expected unique local experts
  = 112 × (1 - (1 - 16/896)^8)
  ≈ 15.0 experts/rank/layer
```

A 200,000-sample uniform placement simulation gives a median maximum of **20 distinct experts on the slowest of eight ranks**, p90 23, p95 24 and p99 26. Real K3 routing is not uniform; hot-expert reuse can reduce unique weight reads, while poor expert placement can increase rank skew. That is why production work must measure actual expert IDs and cannot use the average rank as its roof.

## HBM traffic derivation

One routed K3 expert contains three 3,584×3,072 matrices. Its packed nibbles plus one byte of scale per 32 values occupy **17.55 MB** in the real checkpoint.

| Component per eight-token model call | Derivation | Approx. bytes/GPU/step |
|---|---|---:|
| Fixed MoE router + latent down/up + TP-sharded shared experts | `(12.85 + 102.76 + 33.03) MB × 92` | 13.67 GB |
| KDA + MLA weights | 69×113.7 MB + 24×84.5 MB | 9.87 GB |
| Dense layer, embedding/head, norms and small tensors | model-derived allowance | ~0.7–1.0 GB |
| 4k FP8 MLA KV reads for eight sequences | approximately 24 layers × 4k × latent/rope bytes × 8 | ~0.45 GB |
| Routed experts, average rank | 15.0×17.55 MB×92 | 24.2 GB |
| Routed experts, median slowest rank | 20×17.55 MB×92 | 32.3 GB |
| Routed experts, p95 slowest rank | 24×17.55 MB×92 | 38.7 GB |

This yields approximately **49 GB on an average rank, 57 GB on the median critical rank, and 63–64 GB at p95**. The ideal bandwidth-only roof is therefore:

```text
average rank:        49 GB / 8 GB/ms = 6.1 ms/step => ~1,300 tok/s
median critical:     57 GB / 8 GB/ms = 7.1 ms/step => ~1,120 tok/s
p95 critical:      63.5 GB / 8 GB/ms = 7.9 ms/step => ~1,010 tok/s

aggregate decode rate = 8 output tokens / model-step time
```

K3 performs roughly 1.6 trillion arithmetic operations for an eight-token step. Even pessimistically pricing all of them at the aggregate BF16 peak is about 0.087 ms, far below the 6–8 ms HBM floor. Actual M=8 kernels do not reach peak matrix throughput, but this gap establishes the correct roofline regime: **weight movement, skinny/grouped kernel efficiency, rank imbalance, reductions and launch serialization dominate—not nominal FLOPs**.

The traffic estimates assume each fixed weight tile is fetched once per step. If the current kernels reload a tile for separate tokens or padded rows, measured DRAM bytes will be higher; that inefficiency is part of the opportunity, not part of the theoretical roof.

## Current trace evidence

The real-checkpoint rank-0 decode trace contains **2,690 GPU dispatches per eight-token model call**. The first 64 graph replays report 38.36 ms of summed GPU dispatch time and 40.45 ms per profiled `hipGraphLaunch`, versus 35.77 ms unprofiled TPOT. The profiler perturbs timing, and overlapping/spin-wait durations are not additive critical-path time.

The largest exact kernels per model step are:

| Kernel | Attribution | Summed time/step |
|---|---|---:|
| `iris_reduce_symmetric_gluon_kernel.kd` | MoE reduction/rendezvous | 6.17 ms |
| `_stage1_a16w4_situ_warp_gemv.kd` | Routed expert W13 + SiTU | 6.05 ms |
| `iris_stage_one_shot_allreduce_kernel.kd` | TP collective/residual | 4.05 ms |
| `_stage2_a16w4_warp_gemv_combine.kd` | Routed expert W2/combine | 2.96 ms |
| PyTorch `direct_copy_kernel_cuda` family | copies/layout materialization | 2.56 ms |
| `_packed_projection_gemm_kernel.kd` | latent/output projection | 2.01 ms |

First-principles interpretation:

1. **Routed experts and their rendezvous are the central b8 path.** Stage 1 plus stage 2 is ~9 ms of dispatch work, and the symmetric reduction waits for the slowest producer.
2. **The batch is not dense from an expert's perspective.** Roughly 120 unique experts serve 128 assignments under uniform routing, so most expert GEMVs still see one row.
3. **Fixed matrices should reuse weights eight ways.** Any near-linear growth from b1 indicates padding, reloading, or a schedule that fails to exploit M=8.
4. **2,690 dispatches remain expensive inside a graph.** CUDA/HIP graph replay removes Python/CPU launch overhead, not GPU command processing, kernel tails, or inter-kernel dependencies.
5. **Rank 0 is not necessarily the slowest rank.** Per-layer max expert count and collective arrival skew determine the wall TPOT.

### P0 — turn the roof assumptions into measurements

- Collect rocprofv3 HBM-read/cache-hit counters for stage-1, stage-2, packed projection, KDA/MLA projections and the largest dense GEMMs. If graph replay prevents counters, replay identical isolated shapes eagerly for counters while retaining graph mode for end-to-end timing.
- Record selected expert IDs and unique expert count per layer/rank. Report average, median-max, p95-max and the slowest rank's bytes.
- Produce a non-overlapping critical-path timeline with collective arrival timestamps. Do not equate a 6.17 ms Iris duration with 6.17 ms removable wall time—the kernel can include producer skew and waiting.
- Track useful bytes versus actual HBM bytes. This directly tests whether M=8 fixed weights are fetched once or redundantly.

### P1 — optimize the sparse grouped-expert path for “mostly M=1” experts

- Replace the current warp-GEMV regime with a grouped schedule that assigns work by expert/token occupancy, not a uniform worst-case tile. Most selected experts have one token, but a few have two or more; dispatch tactics by occupancy bucket.
- Keep packed MXFP4 weights/scales contiguous and coalesce expert transitions. A persistent work queue can let CUs pull `(expert, tile)` tasks without leaving waves idle behind a heavily routed rank.
- Fuse stage-1 SiTU epilogues and stage-2 weighted combine without increasing the slowest expert tail. Optimize time-to-last-expert, not mean expert throughput.
- Gate on slowest-rank stage completion and require stage1+stage2 to fall from ~9 ms profiled dispatch toward ≤5 ms before claiming success.

### P2 — make M=8 fixed projections actually reuse their weights

- Measure DRAM bytes for router/routed-down/shared gate-up, routed-up, KDA q/k/v/g, MLA projections and shared down. The ideal is one weight-tile read for eight activation rows.
- Add M=8-specific Gluon tiles rather than reusing b1 warp-GEMV or b16 schedules. Avoid padding useful M=8 work to a tile that doubles row computation or output traffic.
- Preserve the packed router + routed-down + shared-input projection, but compare it against the best standalone M=8 kernels; fusion must improve effective HBM bandwidth and critical-tail time, not only launch count.
- Fuse residual/norm/activation epilogues when that removes full activation round trips.

### P3 — balance EP8 before the reduction

- Reorder or replicate hot experts so each layer minimizes the maximum distinct experts per rank. At b8, reducing the slowest rank from 24 to 20 unique experts saves about 6.5 GB of expert reads across 92 layers.
- Evaluate expert placement on real request distributions. Uniform synthetic routing is useful only as a control.
- Start Iris directly from the producer stream and keep routed/shared partials in the existing joint reduction. Use arrival traces to decide whether the next change belongs in expert scheduling or the collective primitive.
- Treat expert replication as a memory/capacity tradeoff and retain the long-context capacity gates from #55.

### P4 — reduce fixed BF16 traffic

The fixed BF16 stream is ~24 GB/step before experts, so it remains material even though routed-expert bytes grow at b8.

- Prototype W8A16/FP8 weight-only kernels for latent down/up, shared experts, and KDA/MLA projections.
- Benchmark each matrix family independently; at M=8, dequantization and layout conversion can erase the byte saving.
- Keep BF16 accumulation/output first. Require AIME26 and targeted logit agreement before combining families.
- Consider MXFP4 for currently ignored BF16 matrices only after the W8 path establishes accuracy headroom.

### P5 — shrink the 2,690-dispatch graph

- First target: <2,000 dispatches/step. Stretch: <1,400.
- Prioritize repeated per-layer operations: route/quantize, expert occupancy scans, copies around stage1/2, KDA state epilogues, MLA output/residual, and reduction handoffs.
- Replace PyTorch copy/layout materialization with producer-native layouts or fused consumers.
- Reject fusions that reduce launch count but increase registers, lower occupancy, or extend the last producer on the critical rank.

## Targets and validation

| Milestone | Model-step TPOT | Steady decode rate | Approximate objective |
|---|---:|---:|---|
| Current TokenSpeed | 35.77 ms | 223.6 tok/s | 207.05 observed output tok/s |
| Current SGLang reference | 25.68 ms | 311.5 tok/s | 295.31 observed output tok/s |
| Match | ≤26.7 ms | ≥300 tok/s | Close the known competitive gap |
| **Production** | **≤20 ms** | **≥400 tok/s** | About 35–40% of critical-rank HBM roof |
| **Stretch** | **≤16 ms** | **≥500 tok/s** | Requires expert balance + strong M=8 reuse |
| Hardware-only critical-rank roof | 7.1–8.0 ms | 1,000–1,120 tok/s | Perfect streaming; not a release target |

Accept a change only when:

1. The real K3 checkpoint runs TP8/EP8 with decode graphs, FP8 KV, exact 4k/1k and b8/c8.
2. Unprofiled median and p95 TPOT are measured after graph warmup over at least three waves; profiling happens in a separate run.
3. Slowest-rank time, expert occupancy and actual HBM bytes are reported.
4. AIME26 and targeted logits stay within the established tolerance.
5. b1/c1 remains within 2% of its 16.49 ms TPOT, b16/c16 does not regress by more than 2%, and long-context capacity is unchanged.

## Related evidence

- #55 contains the complete TP8/EP8 matrix, per-stage top-kernel tables and visual graph timelines.
- #57 derives the corresponding b1/c1 roofline and explains the fixed-weight-dominated regime.
- #53 tracks productionizing the optimized K3 execution path and earlier Proton reduction analysis.
- Local traces and derivation helpers: `/home/stwinata/k3_sgl_vs_ts_perf/artifacts/matrix_4k_128k_20260813/analysis/`.

The b8 optimization question is therefore: **how do we stream roughly 57–64 GB on the slowest rank, finish ~120 distinct experts across the node, and complete 92 dependent reductions in ≤20 ms—while preserving b1?**
